oshmem/mca/scoll/basic/scoll_basic_alltoall.c

```diff
@@ -96,28 +126,47 @@ static int _algorithm_simple(struct oshmem_group_t *group,
         /* index permutation for better distribution of traffic */
         k = (((j)+(i))%(group->proc_count));
         pe_cur = oshmem_proc_pe(group->proc_array[k]);
-        rc = MCA_SPML_CALL(put(
-                (void *)((char *)target + j * tst * nelems * element_size),
-                nelems * element_size,
-                (void *)((char *)source + i * sst * nelems * element_size),
-                pe_cur));
-        if (OSHMEM_SUCCESS != rc) {
-            break;
+        for (l = 0; l < nelems; l++) {
+            rc = MCA_SPML_CALL(put(
+                        get_stride_elem(target, tst, nelems, element_size, j, l),
+                        element_size,
+                        get_stride_elem(source, sst, nelems, element_size, i, l),
+                        pe_cur));
+            if (OSHMEM_SUCCESS != rc) {
+                return rc;
+            }
         }
     }
-    /* fence (which currently acts as quiet) is needed
-     * because scoll level barrier does not guarantee put completion
-     */
-    MCA_SPML_CALL(fence());
+    return OSHMEM_SUCCESS;
+}
 
-    /* Wait for operation completion */
-    if (rc == OSHMEM_SUCCESS) {
-        SCOLL_VERBOSE(14, "[#%d] Wait for operation completion", group->my_pe);
-        rc = BARRIER_FUNC(group,
-                (pSync + 1),
-                SCOLL_DEFAULT_ALG);
-    }
+static int a2a_alg_simple(struct oshmem_group_t *group,
+                           void *target,
+                           const void *source,
+                           size_t nelems,
+                           size_t element_size)
+{
+    int rc;
+    int pe_cur;
+    int i;
+    int j;
+    int k;
 
-    return rc;
+    SCOLL_VERBOSE(14,
+                  "[#%d] send data to all PE in the group",
+                  group->my_pe);
+    j = oshmem_proc_group_find_id(group, group->my_pe);
+    for (i = 0; i < group->proc_count; i++) {
+        /* index permutation for better distribution of traffic */
+        k = (((j)+(i))%(group->proc_count));
+        pe_cur = oshmem_proc_pe(group->proc_array[k]);
+        rc = MCA_SPML_CALL(put((char *)target + j * nelems * element_size,
```

Comment:
maybe we can reuse and call `get_stride_elem(target, 1, nelems, element_size, j, 0)`

Reply:
also seems like the 1st argument to put() can be pre-calculated before the loop